File: 113-path-sum-ii/path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        lst=[]
        res=[]
        def dfs(root,targetSum):
            if root == None:
                return []
            lst.append(root.val)
            targetSum-=root.val
            if(root.left == None and root.right == None and targetSum==0):
                res.append(lst[:])
            dfs(root.left,targetSum)
            dfs(root.right,targetSum)

            lst.pop()

            return res
        return dfs(root,targetSum)

File: 113-path-sum-ii/path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void recur(TreeNode* root,int targetSum,vector<int>& path,vector<vector<int>>& res)
    {
        if(root == nullptr)
        {
            return;
        }
        path.push_back(root->val);
        if(root->left == nullptr && root->right == nullptr)
        {
            if(targetSum==root->val)
            {
                res.push_back(path);
            }
        }
                targetSum-= root->val;

        recur(root->left,targetSum,path,res);
        recur(root->right,targetSum,path,res);
        path.pop_back();

    }

    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> path;
        vector<vector<int>> res;
        recur(root,targetSum,path,res);
        return res;
    }
};